Coll framework prereqs

wesbland opened the pull request.

wesbland commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

wesbland commented:
Missed a function

test:jenkins/ch4/most
test:jenkins/ch3/most

pavanbalaji commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

wesbland commented on src/mpi/coll/bcast.c:
```diff
@@ -1095,28 +1079,43 @@ static int MPIR_SMP_Bcast(
             /* perform the internode broadcast */
             if (comm_ptr->node_roots_comm != NULL)
             {
                 if (MPIU_is_pof2(comm_ptr->node_roots_comm->local_size, NULL))
```
We can just remove this `if` statement now that both cases do the same thing.

pavanbalaji replied:
Fixed.

wesbland commented on src/mpi/coll/ialltoallw.c:
```diff
@@ -247,6 +247,29 @@ int MPIR_Ialltoallw_inter_sched(const void *sendbuf, const int sendcounts[], con
+int MPIR_Ialltoallw_sched(const void *sendbuf, const int sendcounts[], const int sdispls[],
+                          const MPI_Datatype sendtypes[], void *recvbuf, const int recvcounts[],
+                          const int rdispls[], const MPI_Datatype recvtypes[],
+                          MPIR_Comm *comm_ptr, MPIR_Sched_t s)
```
This ended up in the wrong patch. It should move backward a couple.

wesbland commented on src/mpid/ch3/include/mpid_coll.h:
```diff
@@ -718,3 +718,382 @@ static inline int MPID_Iscatterv(const void *sendbuf, const int *sendcounts,
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    mpi_errno = MPIR_Ibarrier_intra_sched(comm, s);
```
Why are we using the `_intra_` versions of these collectives here? Shouldn't we just call `MPIR_Ibarrier_sched`?

pavanbalaji replied:
This is equivalent to the `coll_fns`.  I didn't want to change that functionality.

pavanbalaji commented:
@wesbland I've force-pushed the changes requested.

wesbland commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

wesbland commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

wesbland commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

wesbland commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

wesbland commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

pavanbalaji commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

wesbland commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

wesbland commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

pavanbalaji commented:
test:jenkins/ch4/most
test:jenkins/ch3/most

pavanbalaji commented:
I'm seeing some warnings in the build:

```
  CC       src/mpi/coll/red_scat_block/lib_libmpi_la-red_scat_block_recursive_doubling.lo
../mpich/src/mpi/coll/red_scat_block/red_scat_block_pairwise.c: In function 'MPIR_Reduce_scatter_block_pairwise':
../mpich/src/mpi/coll/red_scat_block/red_scat_block_pairwise.c:38:9: warning: variable 'total_count' set but not used [-Wunused-but-set-variable]
     int total_count, src, dst;
         ^~~~~~~~~~~
```

This is with `--enable-strict` and gcc-7, though I assume that most compilers will throw that warning with `--enable-strict`.

pavanbalaji commented on src/mpi/coll/collutil.h:
```diff
@@ -7,6 +7,10 @@
+int MPII_Allreduce_group(void *sendbuf, void *recvbuf, int count,
+                         MPI_Datatype datatype, MPI_Op op, MPIR_Comm *comm_ptr,
+                         MPIR_Group *group_ptr, int tag, MPIR_Errflag_t *errflag);
+
```
We should either declare the `MPII_Allreduce_group_intra` function (so other files can use it) or define it as `static` and remove the `MPII_` namespace.

wesbland replied:
Agreed. I chose to add the `MPII_Allreduce_group_intra` function here. That was an oversight and even though we don't use it right now (as seen by the warning below), we shouldn't remove the option.

wesbland commented:
@pavanbalaji - I fixed all of the warnings above.

I've also added a patch with a preview of what the manual CVAR algorithm selection will look like in fa3c1fd. I'll keep working on the rest of the algorithms, but if you don't like what you see in that one, I should change the tactic for the rest of them.

pavanbalaji commented on src/mpi/coll/bcast/bcast.c:
```diff
@@ -581,9 +606,17 @@ int MPIR_Bcast(void *buffer, int count, MPI_Datatype datatype, int root, MPIR_Co
-        mpi_errno = MPIR_Bcast_intra( buffer, count, datatype, root, comm_ptr, errflag );
+        switch (MPIR_Bcast_alg_intra_choice) {
+            case MPIR_BCAST_ALG_INTRA_BINOMIAL:
+                mpi_errno = MPIR_Bcast_binomial(buffer, count, datatype, root, comm_ptr, errflag);
```
You need appropriate breaks in here.

wesbland replied:
Already done.

Bayyapu commented:
Inter and Intra Communicator functions are missing in Exscan and Gatherv functions.

Bayyapu commented:
@wesbland  The following files need updates:
intra and inter functions are not defined for the following functions:
exscan
gatherv
iexscan
igatherv
inhb_allgather
inhb_allgatherv
inhb_alltoall
inhb_alltoallv
inhb_alltoallw
iscan
iscatter
nhb_allgather
nhb_allgatherv
nhb_alltoall
nhb_alltoallv
nhb_alltoallw
scan
scatterv
Further, allreduce_group and barrier_group related files do not exist.

wesbland commented on src/mpi/coll/src/coll_impl.c:
```diff
@@ -52,32 +52,78 @@ int MPIR_COLL_init(void) {
 
     /* Allreduce_group Intra */
 
     /* Allreduce_group Inter */
```
I know some of the collectives don't have alternative implementations (allred_group, exscan, gatherv, neighborhood collectives, etc.), but I think everything should get a CVAR.

wesbland replied:
If we're not going to do allreduce_group (and any others that we're not going to do, let's take them out of this list.

Bayyapu replied:
@wesbland In Allgatherv Inter above, As there is only one option we have not used string comparison. Should this also be changed?

wesbland replied:
Yes, sorry I missed that one.

Bayyapu replied:
@wesbland Ok, I have a to add a good number of changes then. Will do and update Thank you

wesbland commented on src/mpi/coll/allgatherv/allgatherv.c:
```diff
@@ -22,6 +22,45 @@
+        If set to true, MPI_Allgatherv will allow the device to override the
+        default, MPIR-level collective algorithms. The device still has the
+        option to call the MPIR-level algorithms manually.
+        If set to false, the device-level allgather function will not be
```
This needs to get fixed for all collectives. The comment still says `allgather`.

wesbland commented on src/mpi/coll/allgatherv/allgatherv.c:
```diff
@@ -299,17 +338,40 @@ int MPIR_Allgatherv(const void *sendbuf, int sendcount, MPI_Datatype sendtype,
     } else {
-        /* intracommunicator */
+        switch (MPIR_Allgatherv_alg_inter_choice) {
+            default:
```
I know I missed it above, but I'd like to see `case MPIR_ALLGATHERV_ALG_INTER_AUTO:` handled before the `default:` case here just in case we ever have other algorithms.

Obviously, that applies to all of the collectives too.

wesbland commented on src/mpi/coll/iallgather/iallgather.c:
```diff
@@ -6,6 +6,76 @@
+      description : >-
+        For MPI_Iallgather and MPI_Iallgatherv, the long message algorithm will be
+        used if the send buffer size is >= this value (in bytes)
+        (See also: MPIR_CVAR_IALLGATHER_SHORT_MSG_SIZE)
```
I'm not sure these two CVARs are necessary. You're not using them below.

wesbland commented on src/mpi/coll/iallgatherv/iallgatherv.c:
```diff
@@ -6,6 +6,76 @@
+      description : >-
+        For MPI_Iallgather and MPI_Iallgatherv, the long message algorithm will be
+        used if the send buffer size is >= this value (in bytes)
+        (See also: MPIR_CVAR_IALLGATHERV_SHORT_MSG_SIZE)
```
Same as above, these CVARs are unnecessary.

wesbland commented on src/mpi/coll/iallreduce/iallreduce.c:
```diff
@@ -245,14 +291,37 @@ int MPIR_Iallreduce_sched(const void *sendbuf, void *recvbuf, int count, MPI_Dat
-    return mpi_errno;
 }
+    return mpi_errno;
+
```
This looks weird. Did it get moved accidentally?

wesbland commented on src/mpi/coll/ialltoallv/ialltoallv.c:
```diff
@@ -89,14 +135,37 @@ int MPIR_Ialltoallv_sched(const void *sendbuf, const int sendcounts[], const int
+                break;
+        }
     }
-
```
Please revert the whitespace changes.

wesbland commented on src/mpi/coll/ialltoallv/ialltoallv.c:
```diff
@@ -89,14 +135,37 @@ int MPIR_Ialltoallv_sched(const void *sendbuf, const int sendcounts[], const int
-
     return mpi_errno;
 }
-
```
Please revert the whitespace changes.

wesbland commented on src/mpi/coll/ialltoallw/ialltoallw.c:
```diff
@@ -106,15 +152,35 @@ int MPIR_Ialltoallw_sched(const void *sendbuf, const int sendcounts[], const int
+                break;
+        }
     }
-
```
Please revert the whitespace changes.

wesbland commented on src/mpi/coll/ibarrier/ibarrier.c:
```diff
@@ -86,11 +131,28 @@ int MPIR_Ibarrier_sched(MPIR_Comm *comm_ptr, MPIR_Sched_t s)
+                break;
+        }
     }
-
```
Please revert the whitespace changes.

wesbland commented on src/mpi/coll/ibcast/ibcast.c:
```diff
@@ -8,6 +8,131 @@
+      description : >-
+        Maximum message size for which SMP-aware broadcast is used.  A
+        value of '0' uses SMP-aware broadcast for all message sizes.
+        (See also: MPIR_CVAR_ENABLE_SMP_IBCAST)
```
I think all of these CVARs can go and we can use the regular bcast CVARs.

wesbland commented on src/mpi/coll/ibcast/ibcast.c:
```diff
@@ -217,14 +342,35 @@ int MPIR_Ibcast_sched(void *buffer, int count, MPI_Datatype datatype, int root,
     if (comm_ptr->comm_kind == MPIR_COMM_KIND__INTRACOMM) {
         if (comm_ptr->hierarchy_kind == MPIR_COMM_HIERARCHY_KIND__PARENT) {
-            mpi_errno = ibcast_smp_sched(buffer, count, datatype, root, comm_ptr, s);
+                mpi_errno = ibcast_smp_sched(buffer, count, datatype, root, comm_ptr, s);
```
Not sure why the extra indentation is here.

wesbland commented on src/mpi/coll/ibcast/ibcast.c:
```diff
@@ -217,14 +342,35 @@ int MPIR_Ibcast_sched(void *buffer, int count, MPI_Datatype datatype, int root,
+                break;
+         }
     }
-
```
Please revert the whitespace changes.

wesbland commented on src/mpi/coll/ired_scat/ired_scat.c:
```diff
@@ -7,6 +7,65 @@
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        the long message algorithm will be used if the operation is commutative
+        and the send buffer size is >= this value (in bytes)
```
This CVAR is unnecessary.

wesbland commented on src/mpi/coll/ired_scat_block/ired_scat_block.c:
```diff
@@ -7,6 +7,65 @@
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        the long message algorithm will be used if the operation is commutative
+        and the send buffer size is >= this value (in bytes)
```
Unnecessary CVAR.

wesbland commented on src/mpi/coll/ired_scat_block/ired_scat_block.c:
```diff
@@ -102,11 +161,42 @@ int MPIR_Ireduce_scatter_block_sched(const void *sendbuf, void *recvbuf, int rec
+                           recvcounts, datatype, op, comm_ptr, s);
+       }
     }
-
```
Please revert the whitespace changes.

wesbland commented on src/mpi/coll/ireduce/ireduce.c:
```diff
@@ -7,6 +7,72 @@
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        Enable SMP aware ireduce.
```
Unnecessary CVARs.

wesbland commented on src/mpi/coll/red_scat_block/red_scat_block.c:
```diff
@@ -14,6 +14,64 @@
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        the long message algorithm will be used if the operation is commutative
+        and the send buffer size is >= this value (in bytes)
```
Unnecessary CVAR

pavanbalaji commented:
@wesbland I don't understand your comment about removing some CVARs.  The message or comm size based CVARs are still needed when you use the auto-selection.

pavanbalaji commented:
For some collectives, e.g., `gatherv`, intercomms are definitely valid.  If we didn't implement them, they should be implemented.

wesbland commented:
@pavanbalaji The CVARs duplicate those that already exist for the blocking versions of those algorithms. From what I can tell, the new CVARs aren't used anywhere, but were just copied from the blocking versions.

I agree with what you said about intercomms for collectives where they make sense. In those cases, we should cover both cases but have them both use the same call for the `_AUTO` case.

pavanbalaji commented:
@wesbland Oh, I didn't realize those were duplicated for the nonblocking collectives.  Yes, those should be removed.

For the intercomm collectives, some of the algorithms are wrong.  I believe the `gatherv` algorithm will fail for intercomms, but someone should verify that.

wesbland commented:
I just think we should use whatever algorithm for whatever case we were using before, but wrap them in CVARs in case we ever add more.

pavanbalaji commented:
I agree that this patch should not be anything more than adding CVARs.  We should not make algorithmic changes here.  I was suggesting doing a separate patch that fixes the issue with intercommunicators.  @Bayyapu do you think you can fix those up as a separate commit (and add a test case for it)?

pavanbalaji commented:
@Bayyapu your commits have created a mess in the branch.  There are conflict markups, one of Wesley's commits is duplicated.  What's going on?

Bayyapu commented:
@pavanbalaji and @wesbland I will remove those CVARs checks according to the comments

wesbland commented:
It looks like this was squashed with the wrong commit. I have an old version of this branch from a few hours ago that I can force push if it helps.

wesbland commented on src/mpi/coll/red_scat/red_scat.c:
```diff
@@ -12,16 +12,45 @@
 === BEGIN_MPI_T_CVAR_INFO_BLOCK ===
 
 cvars:
-    - name        : MPIR_CVAR_REDSCAT_COMMUTATIVE_LONG_MSG_SIZE
```
Why did this CVAR go away?

Bayyapu replied:
@wesbland this should be there and will insert it back

wesbland commented on src/mpi/coll/src/coll_impl.c:
```diff
@@ -101,52 +147,141 @@ int MPIR_COLL_init(void) {
     /* Gather Inter */
+    MPIR_Gather_alg_inter_choice = MPIR_GATHER_ALG_INTER_AUTO;
 
     /* Gatherv Intra */
```
We're still missing a bunch of collectives in this list. Even if we only have the `_AUTO` options, I think we should still put the infrastructure in here to do the algorithm selection.

Bayyapu replied:
@wesbland There are no inter and intra functions defined within gatherv and other functions. Is it ok to define them in coll_impl.c?

wesbland replied:
Do the same things we do for the other functions, but then have them both call the same function.

Bayyapu replied:
@wesbland Similarly allreduce_group has only intra and inter is not defined.

wesbland replied:
Hmm. Is it illegal to do an allreduce_group on an intercommunicator? I'm not sure what the right answer is. @pavanbalaji?

wesbland replied:
I think it should look the same as the others, string comparisons and all. You'd create the CVARs in `gather.c`, but there would only be one option (the same way we only have one option for most intercommunicator collectives).

Bayyapu replied:
@wesbland For example, In Allgatherv Inter in coll_impl.c, As there is only one option we have not used string comparison. Should this also be changed?

pavanbalaji replied:
@wesbland allreduce_group is an internal function that is only used on intracommunicators.

Bayyapu replied:
@wesbland  May I request you to give a pass on the following:
    /* Scatterv Intra */
    if (0 == strcmp(MPIR_CVAR_SCATTERV_ALGORITHM_INTRA, "linear"))
        MPIR_Scatterv_alg_intra_choice = MPIR_SCATTERV_ALG_LINEAR;
    else
        MPIR_Scatterv_alg_intra_choice = MPIR_SCATTERV_ALG_AUTO;

    /* Scatterv Inter */
    if (0 == strcmp(MPIR_CVAR_SCATTERV_ALGORITHM_INTER, "linear"))
        MPIR_Scatterv_alg_inter_choice = MPIR_SCATTERV_ALG_LINEAR;
    else
        MPIR_Scatterv_alg_inter_choice = MPIR_SCATTERV_ALG_AUTO;

wesbland replied:
@pavanbalaji - Ok, then let's drop the intercommunicators here.

@Bayyapu - Yep, that looks right to me.

pavanbalaji replied:
For the collectives where intercommunicators are not allowed, we should throw an error instead of simply pointing to the same function (in a separate patch).

pavanbalaji commented:
@Bayyapu @wesbland Intercommunicators are not allowed for `scan`, `exscan`, `iscan` and `iexscan`.  We should throw `MPI_ERR_ARG` in case the user passes an intercomm to these functions.

@Bayyapu The `gatherv_linear` function works for both `intra` and `inter` communicators.  So you should create `intra` and `inter` versions that both internally call that function.  Same thing for `igatherv`.

@wesbland did we accidentally forget to implement `inter` communicators for `iscatter`?

@wesbland there is some redundant `scan_smp` code in the `scan` directory.  Can you clean that up?

@wesbland did you find out anything about whether neighborhood collectives are valid for intercommunicators?  My guess is that we simply forgot about them while defining those collectives.  If that's true, we should explicitly point that out in the standard (at least as an errata for now).

@Bayyapu does the above list address all of the functions with missing intra/inter that you found or did I miss something?

Bayyapu commented:
@pavanbalaji One more is scatterv, but anyway you mentioned iscatterv. So all are covered. I will add the inter function for these two. Other suggestions too, I will incorporate accordingly.

pavanbalaji commented:
@Bayyapu I've fixed my above comment to include `scatterv` and `iscatterv` now.

wesbland commented:
> @Bayyapu @wesbland Intercommunicators are not allowed for scan, exscan, iscan and iexscan. We should throw MPI_ERR_ARG in case the user passes an intercomm to these functions.

Do we need to do that in the `MPIR` functions? I assume that's already happening in the `MPI` functions themselves.

> @wesbland did we accidentally forget to implement inter communicators for iscatter?

Not sure what you mean. It's in iscatter.v (https://github.com/wesbland/mpich/blob/af3285862366232c1374e6d04581a22a92093e38/src/mpi/coll/iscatter/iscatter.c#L270).

> @wesbland there is some redundant scan_smp code in the scan directory. Can you clean that up?

Sure.

> @wesbland did you find out anything about whether neighborhood collectives are valid for intercommunicators? My guess is that we simply forgot about them while defining those collectives. If that's true, we should explicitly point that out in the standard (at least as an errata for now).

I saw that OMPI also excludes them and stopped. Since I'm at the forum this week, I can ask people.

pavanbalaji commented:
>> @Bayyapu @wesbland Intercommunicators are not allowed for scan, exscan, iscan and iexscan. We should throw MPI_ERR_ARG in case the user passes an intercomm to these functions.

> Do we need to do that in the MPIR functions? I assume that's already happening in the MPI functions themselves.

Oh, yes, you are right.  The MPI function is checking for it in `MPIR_ERRTEST_COMM_INTRA`.

wesbland commented:
I cleaned up the `scan_smp` thing and added another patch for moving utarray into MPL. Note that this is a force push so please make sure not to blow my changes away. 😄

pavanbalaji commented:
@wesbland we had removed Neelima's patch because it was not ready.  Your force push added that back.  Can you please remove it?

wesbland commented:
Done. Sorry about that.

wesbland commented:
@pavanbalaji - I've updated this PR with the current status. I'm still working on a few things:

* There's a few algorithms with `default` in their name. They need to be changed to something else per previous discussions.
* Some of the CVARs have their names inverted (`_INTER` at the end instead of the middle).
* Still working on neighborhood collectives.

There's a bunch of other new patches that you can take a look at though. I've verified that these compile, but I have no idea if they actually run. 🤞 

test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
@wesbland The non-static functions need to be protected inside `#ifndef MPICH_MPI_FROM_PMPI`.  OSX is failing because of this.

@raffenet The UCX failure seems to be something to do with the embedded build.  Is this working on `master`?

pavanbalaji commented on src/mpl/include/mpl_base.h:
```diff
@@ -56,6 +56,16 @@
+#define MPL_FALLTHROUGH ATTRIBUTE((fallthrough))
+#else
+#define MPL_FALLTHROUGH
+#endif
```
This should be a configure check, not just a check for `__GNUC__`.

wesbland replied:
Ah. I finally found the check for this in the autoconf archive. Removed this bandaid.

wesbland commented:
> @wesbland The non-static functions need to be protected inside #ifndef MPICH_MPI_FROM_PMPI. OSX is failing because of this.

I wasn't aware this had to also get done for non-static variables in headers. I guess I didn't think about it. I think I've fixed this now, but I'd rather use Jenkins than set up a build on my laptop.

The only thing left on my list now is to add CVARs for the neighborhood collectives.

test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
I was referring to the functions in the `.c` files, not header files.

pavanbalaji commented:
Headers should not have any direct variable declarations anyway, just externs.

raffenet commented:
Embedded ucx builds are broken in master.

wesbland commented:
Yeah, you're right. These need to move to the .c files... That's the problem.

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

wesbland commented:
Someone with more insight into how the PMPI symbols are generated on MacOS, I'd love some help here. I thought I was doing all the right stuff to avoid the linking errors, but I still see them in my local build:

```
Undefined symbols for architecture x86_64:
  "_MPIR_Allgather_alg_intra_choice", referenced from:
      _MPIR_Allgather in libpmpi.a(lib_libpmpi_la-allgather.o)
  "_MPIR_Allreduce_alg_intra_choice", referenced from:
      _MPIR_Allreduce in libpmpi.a(lib_libpmpi_la-allreduce.o)
  "_MPIR_Alltoall_alg_intra_choice", referenced from:
      _MPIR_Alltoall in libpmpi.a(lib_libpmpi_la-alltoall.o)
  "_MPIR_Alltoallw_alg_intra_choice", referenced from:
      _MPIR_Alltoallw in libpmpi.a(lib_libpmpi_la-alltoallw.o)
  "_MPIR_Barrier_alg_intra_choice", referenced from:
      _MPIR_Barrier_intra in libpmpi.a(lib_libpmpi_la-barrier.o)
      _MPIR_Barrier_inter in libpmpi.a(lib_libpmpi_la-barrier.o)
      _MPIR_Barrier in libpmpi.a(lib_libpmpi_la-barrier.o)
      _PMPI_Barrier in libpmpi.a(lib_libpmpi_la-barrier.o)
  "_MPIR_Bcast_alg_intra_choice", referenced from:
      _MPIR_Bcast in libpmpi.a(lib_libpmpi_la-bcast.o)
  "_MPIR_Gather_alg_intra_choice", referenced from:
      _MPIR_Gather in libpmpi.a(lib_libpmpi_la-gather.o)
  "_MPIR_Iallgather_alg_intra_choice", referenced from:
      _MPIR_Iallgather_sched in libpmpi.a(lib_libpmpi_la-iallgather.o)
  "_MPIR_Iallreduce_alg_intra_choice", referenced from:
      _MPIR_Iallreduce_sched in libpmpi.a(lib_libpmpi_la-iallreduce.o)
  "_MPIR_Ialltoall_alg_intra_choice", referenced from:
      _MPIR_Ialltoall_sched in libpmpi.a(lib_libpmpi_la-ialltoall.o)
  "_MPIR_Ibcast_alg_intra_choice", referenced from:
      _MPIR_Ibcast_sched in libpmpi.a(lib_libpmpi_la-ibcast.o)
  "_MPIR_Igather_alg_inter_choice", referenced from:
      _MPIR_Igather_sched in libpmpi.a(lib_libpmpi_la-igather.o)
  "_MPIR_Igather_alg_intra_choice", referenced from:
      _MPIR_Igather_sched in libpmpi.a(lib_libpmpi_la-igather.o)
  "_MPIR_Ireduce_alg_intra_choice", referenced from:
      _MPIR_Ireduce_sched in libpmpi.a(lib_libpmpi_la-ireduce.o)
  "_MPIR_Reduce_alg_intra_choice", referenced from:
      _MPIR_Reduce in libpmpi.a(lib_libpmpi_la-reduce.o)
  "_MPIR_Reduce_scatter_block_alg_intra_choice", referenced from:
      _MPIR_Reduce_scatter_block in libpmpi.a(lib_libpmpi_la-red_scat_block.o)
  "_MPIR_Scatter_alg_intra_choice", referenced from:
      _MPIR_Scatter in libpmpi.a(lib_libpmpi_la-scatter.o)
```

I don't know why these are showing up for just some of the symbols, but not others. For instance, `MPIR_Allgather_alg_intra_choice` but not `MPIR_Allgather_alg_inter_choice`.

test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
I'll take a look at it.

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

wesbland commented:
MacOS builds are looking good now 👍

pavanbalaji commented:
Why are the `inter` algorithms not split out into separate functions?

wesbland commented:
Ugh. You’re right. I’ll fix that.

pavanbalaji commented:
The algorithm comments are misplaced too.  I’m working on fixing them.

pavanbalaji commented:
Why is `scatter_for_bcast` duplicated across two different files?  It should go into a header file, like `bcast.h` or similar, as a static inline function.  Alternatively, it can be defined in some common `.c` file as an `MPII_` function and declared in a common header file.

wesbland commented:
Yeah, I have no idea how that happened. We do the right thing for the nonblocking version.

wesbland commented:
I think this addresses your concerns @pavanbalaji 

test:jenkins/ch3/most
test:jenkins/ch4/most

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

wesbland commented:
So the tests are looking good. I'm going to hold off on squashing the fixup commits until you think they look fine @pavanbalaji. That way I don't have to dig them back out if there's a problem.

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented on src/include/mpir_coll.h:
```diff
@@ -426,8 +433,10 @@ int MPIR_Iallreduce_sched(const void *sendbuf, void *recvbuf, int count, MPI_Dat
+int MPIR_Iallreduce_intra_smp_sched(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype, MPI_Op op, MPIR_Comm *comm_ptr, MPIR_Sched_t s);
 int MPIR_Ireduce_scatter_sched(const void *sendbuf, void *recvbuf, const int *recvcounts, MPI_Datatype datatype, MPI_Op op, MPIR_Comm *comm_ptr, MPIR_Sched_t s);
 int MPIR_Ireduce_scatter_block_sched(const void *sendbuf, void *recvbuf, int recvcount, MPI_Datatype datatype, MPI_Op op, MPIR_Comm *comm_ptr, MPIR_Sched_t s);
+int MPIR_Ireduce_scatter_block_sched(const void *sendbuf, void *recvbuf, int recvcount, MPI_Datatype datatype, MPI_Op op, MPIR_Comm *comm_ptr, MPIR_Sched_t s);
```
Why is this duplicated?

wesbland replied:
This is an excellent question.

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

wesbland commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
The `ch3:tcp` tests are failing on this branch, apparently with the `aligned_alloc` patch.  How did that patch get through?  Was it not tested thoroughly?

I'll push my signoffs in a bit.

pavanbalaji commented:
To be clear, this needs to be retested after the `aligned_alloc` piece is fixed.

wesbland commented:
Can you stick a revert patch in here to test so we can move forward?

raffenet commented:
The `aligned_alloc` failures are my fault. I only tested on ch4, since that is where the code is used. Should have done more comprehensive... Hajime has submitted fixes that I will review now.

raffenet commented:
Rebased on the `aligned_alloc` fix. Re-testing...
test:jenkins/ch3/most
test:jenkins/ch4/most

raffenet commented:
Test results are good. All commits are signed off. Merging...
